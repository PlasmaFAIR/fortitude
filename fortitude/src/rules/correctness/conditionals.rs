use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::{find_newline, LineEnding, SourceFile};
use ruff_text_size::TextSize;
use tree_sitter::Node;

/// ## What does it do?
/// Checks for misleading semicolons in `if` statements.
///
/// ## Why is this bad?
/// The following code may appear to execute two statements only if the `if`
/// condition is true, but in actuality the second statement will always be
/// executed:
///
/// ```f90
/// if (condition) print *, "Hello"; print *, "World"
/// ```
///
/// It is equivalent to:
///
/// ```f90
/// if (condition) print *, "Hello"
/// print *, "World"
/// ```
///
/// Users should be cautious applying this fix. If the intent was to have
/// both statements execute only if the condition is true, then the user
/// should rewrite the code to use an `if` statement with a block:
///
/// ```f90
/// if (condition) then
///     print *, "Hello"
///     print *, "World"
/// end if
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MisleadingInlineIfSemicolon {}

impl AlwaysFixableViolation for MisleadingInlineIfSemicolon {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Semicolon following inline if-statement is misleading".to_string()
    }

    fn fix_title(&self) -> String {
        "Replace with newline, or convert to `if(...) then` statement".to_string()
    }
}
impl AstRule for MisleadingInlineIfSemicolon {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // If this is an `if (...) then` construct, exit early
        if !inline_if_statement(node) {
            return None;
        }

        // Check that the if statement is followed directly by a semicolon
        if let Some(semicolon_node) = node.next_sibling().filter(|n| n.kind() == ";") {
            // Find the first following non-semicolon node, and check that it's
            // on the same line as the first semicolon node.
            // Comments on the same line are permitted.
            let mut current_node = semicolon_node;
            while let Some(next) = current_node.next_sibling() {
                if next.kind() == ";" {
                    current_node = next;
                    continue;
                }
                if next.start_position().row != semicolon_node.start_position().row
                    || next.kind() == "comment"
                {
                    return None;
                }
                break;
            }
            // Replace semicolon node with a newline to make it clear that the
            // second statement is not part of the if statement. We should also
            // take care to handle indentation and other whitespace.
            let start = semicolon_node.start_textsize();
            let mut end = semicolon_node.end_byte();
            let text = src.source_text().as_bytes();
            while text[end] == b' ' || text[end] == b'\t' {
                end += 1;
            }
            let end = TextSize::try_from(end).unwrap();
            let indentation = node.indentation(src);
            let edit = Edit::replacement(format!("\n{indentation}"), start, end);
            return some_vec!(Diagnostic::from_node(
                MisleadingInlineIfSemicolon {},
                &semicolon_node
            )
            .with_fix(Fix::safe_edit(edit)));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["if_statement"]
    }
}

/// ## What does it do?
/// Checks for misleading line continuations in inline `if` statements.
///
/// ## Why is this bad?
/// An inline `if` statement followed immediately by a line continuation
/// can be easily confused for a block `if` statement:
///
/// ```f90
/// if (condition) &
///     call a_very_long_subroutine_name(with, many, ..., arguments)
/// ```
///
/// If a developer wishes to add a second statement to the `if` 'block',
/// they may be tempted to write:
///
/// ```f90
/// if (condition) &
///     call a_very_long_subroutine_name(with, many, ..., arguments)
///     call another_subroutine(args, ...)  ! Always executes!
/// ```
///
/// To avoid this confusion, inline `if` statements that spill over multiple
/// lines should be written as an if-then-block:
///
/// ```f90
/// if (condition) then
///     call a_very_long_subroutine_name(with, many, ..., arguments)
///     call another_subroutine(args, ...)  ! Only executes if condition is true
/// end if
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MisleadingInlineIfContinuation {}

impl AlwaysFixableViolation for MisleadingInlineIfContinuation {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Line continuation in inline if-statement is misleading".to_string()
    }

    fn fix_title(&self) -> String {
        "Convert to `if(...) then` block".to_string()
    }
}
impl AstRule for MisleadingInlineIfContinuation {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // If this is an `if (...) then` construct, exit early
        if !inline_if_statement(node) {
            return None;
        }

        // Check if the condition is immediately following by a continuation
        let body_start = node
            .child_with_name("parenthesized_expression")?
            .next_sibling()?;
        if body_start.kind() == "&" {
            let content = ifthenify(node, src)?;
            let start_byte = node.start_textsize();
            let end_byte = node.end_textsize();
            let edit = Edit::replacement(content, start_byte, end_byte);
            return some_vec!(
                Diagnostic::from_node(MisleadingInlineIfContinuation {}, node)
                    .with_fix(Fix::safe_edit(edit))
            );
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["if_statement"]
    }
}

/// Given an if statement, return true if it is inline.
pub fn inline_if_statement(node: &Node) -> bool {
    node.child_with_name("end_if_statement").is_none()
}

/// Given an if statement node, convert it from a one-line if statement
/// to a block if statement.
/// Returns None if the node is already a block if statement.
pub fn ifthenify(node: &Node, src: &SourceFile) -> Option<String> {
    // check that the node is an if statement without an end if statement
    if !inline_if_statement(node) {
        return None;
    }
    let text = node.to_text(src.source_text())?.trim();
    let bytes = text.as_bytes();
    // If CrLf line endings are used, be careful to keep them in the fix
    let nl = find_newline(text)
        .map(|(_, ending)| ending)
        .unwrap_or(LineEnding::Lf)
        .as_str();
    // Divide the if statement into everything up to the end of the condition
    // and the body.
    let condition_node = node.child_with_name("parenthesized_expression")?;
    let condition_end_byte = condition_node.end_byte() - node.start_byte();
    let mut body_start_byte = condition_end_byte;
    // Skip over any whitespace or line continuations between the condition
    // and the body.
    while body_start_byte < node.end_byte()
        && (bytes[body_start_byte].is_ascii_whitespace() || bytes[body_start_byte] == b'&')
    {
        body_start_byte += 1;
    }
    // Build the new if statement.
    let prelude = &text[..condition_end_byte];
    let body = &text[body_start_byte..];
    let indentation = node.indentation(src);
    // Assume two-space indent for the if body
    Some(format!(
        "{} then{nl}{indentation}  {}{nl}{indentation}end if",
        prelude, body
    ))
}
