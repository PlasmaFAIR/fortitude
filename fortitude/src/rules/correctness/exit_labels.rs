use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// When using `exit` or `cycle` in a named `do` loop, the `exit`/`cycle` statement
/// should use the loop name
///
/// ## Example
/// ```f90
/// name: do
///   exit name
/// end do name
/// ```
///
/// Using named loops is particularly useful for nested or complicated loops, as it
/// helps the reader keep track of the flow of logic. It's also the only way to `exit`
/// or `cycle` outer loops from within inner ones.
#[derive(ViolationMetadata)]
pub(crate) struct MissingExitOrCycleLabel {
    name: String,
    label: String,
}

impl Violation for MissingExitOrCycleLabel {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name, label } = self;
        format!("'{name}' statement in named 'do' loop missing label '{label}'")
    }

    fn fix_title(&self) -> Option<String> {
        let Self { label, .. } = self;
        Some(format!("Add label '{label}'"))
    }
}
impl AstRule for MissingExitOrCycleLabel {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Skip unlabelled loops
        let label = node
            .child_with_name("block_label_start_expression")?
            .to_text(src)?
            .trim_end_matches(':');

        let violations: Vec<Diagnostic> = node
            .named_descendants_except(["do_loop_statement"])
            .filter(|node| node.kind() == "keyword_statement")
            .map(|stmt| (stmt, stmt.to_text(src).unwrap_or_default().to_lowercase()))
            .filter(|(_, name)| name == "exit" || name == "cycle")
            .map(|(stmt, name)| {
                let label_with_space = format!(" {label}");
                let edit = Edit::insertion(label_with_space, stmt.end_textsize());
                let fix = Fix::unsafe_edit(edit);

                Diagnostic::from_node(
                    Self {
                        name: name.to_string(),
                        label: label.to_string(),
                    },
                    &stmt,
                )
                .with_fix(fix)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["do_loop_statement"]
    }
}

/// ## What does it do?
/// Checks for `exit` or `cycle` in unnamed `do` loops
///
/// ## Why is this bad?
/// Using loop labels with `exit` and `cycle` statements prevents bugs when exiting the
/// wrong loop, and helps readability in deeply nested or long loops. The danger is
/// particularly enhanced when code is refactored to add further loops.
///
/// ## Settings
/// See [allow-unnested-loops](../settings.md#check_exit-unlabelled-loops_allow-unnested-loops)
#[derive(ViolationMetadata)]
pub(crate) struct ExitOrCycleInUnlabelledLoop {
    name: String,
}

impl Violation for ExitOrCycleInUnlabelledLoop {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' statement in unlabelled 'do' loop")
    }
}

impl AstRule for ExitOrCycleInUnlabelledLoop {
    fn check(settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = source.source_text();
        let name = node.to_text(src)?.to_lowercase();
        // This filters to the keywords we want that _also_ don't have a label
        if !matches!(name.as_str(), "exit" | "cycle") {
            return None;
        }

        let parent_loop = node
            .ancestors()
            .filter(|ancestor| ancestor.kind() == "do_loop_statement")
            .nth(0)?;

        // Immediate parent loop has a label, but we don't want to warn here, because
        // that's covered by missing-exit-or-cycle-label
        if parent_loop
            .child_with_name("block_label_start_expression")
            .is_some()
        {
            return None;
        }

        // If we're only supposed to check on nested loops, check that there is at least
        // one more level of nesting
        if settings.check.exit_unlabelled_loops.allow_unnested_loops {
            parent_loop
                .ancestors()
                .filter(|ancestor| ancestor.kind() == "do_loop_statement")
                .nth(0)?;
        }

        some_vec!(Diagnostic::from_node(Self { name }, node))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["keyword_statement"]
    }
}

/// ## What does it do?
/// Check for missing block labels on `end` statements
///
/// ## Why is this bad?
/// This is a compile error!
#[derive(ViolationMetadata)]
pub(crate) struct MissingEndLabel {
    start_name: String,
    end_name: String,
    label: String,
}

impl AlwaysFixableViolation for MissingEndLabel {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self {
            start_name,
            end_name,
            ..
        } = self;
        format!("'{end_name}' statement in named '{start_name}' block missing label")
    }

    fn fix_title(&self) -> String {
        let Self { label, .. } = self;
        format!("Add label '{label}'")
    }
}

impl AstRule for MissingEndLabel {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Skip unlabelled loops
        let label = node
            .child_with_name("block_label_start_expression")?
            .to_text(src)?
            .trim_end_matches(':');

        let end = match node.kind() {
            "select_case_statement" | "select_type_statement" | "select_rank_statement" => {
                "end_select_statement".to_string()
            }
            _ => format!("end_{}", node.kind()),
        };

        let violation: Diagnostic = node
            .child_with_name(&end)
            .filter(|node| node.child_with_name("block_label").is_none())
            .map(|stmt| (stmt, stmt.to_text(src).unwrap_or_default().to_lowercase()))
            .map(|(stmt, end_name)| {
                let label_with_space = format!(" {label}");
                let edit = Edit::insertion(label_with_space, stmt.end_textsize());
                let fix = Fix::safe_edit(edit);

                let start_name = node.child(1).unwrap().to_text(src).unwrap().to_lowercase();

                Diagnostic::from_node(
                    Self {
                        end_name: end_name.to_string(),
                        start_name,
                        label: label.to_string(),
                    },
                    &stmt,
                )
                .with_fix(fix)
            })?;

        Some(vec![violation])
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "associate_statement",
            "block_construct",
            "coarray_critical_statement",
            "coarray_team_statement",
            "do_loop_statement",
            "forall_statement",
            "if_statement",
            "select_case_statement",
            "select_rank_statement",
            "select_type_statement",
            "where_statement",
        ]
    }
}

pub(crate) mod settings {
    use crate::display_settings;
    use ruff_macros::CacheKey;
    use std::fmt::{Display, Formatter};

    #[derive(Debug, Clone, Default, CacheKey)]
    pub struct Settings {
        pub allow_unnested_loops: bool,
    }

    impl Display for Settings {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            display_settings! {
                formatter = f,
                namespace = "check.exit_unlabelled_loops",
                fields = [self.allow_unnested_loops]
            }
            Ok(())
        }
    }
}
