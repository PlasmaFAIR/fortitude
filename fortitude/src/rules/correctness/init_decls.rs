use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What it does
/// Checks for local variables with implicit `save`
///
/// ## Why is this bad?
/// Initialising procedure local variables in their declaration gives them an
/// implicit `save` attribute: the initialisation is only done on the first call
/// to the procedure, and the variable retains its value on exit.
///
/// ## Examples
/// For example, this subroutine:
///
/// ```f90
/// subroutine example()
///   integer :: var = 1
///   print*, var
///   var = var + 1
/// end subroutine example
/// ```
///
/// when called twice:
///
/// ```f90
/// call example()
/// call example()
/// ```
///
/// prints `1 2`, when it might be expected to print `1 1`.
///
/// Adding the `save` attribute makes it clear that this is the intention:
///
/// ```f90
/// subroutine example()
///   integer, save :: var = 1
///   print*, var
///   var = var + 1
/// end subroutine example
/// ```
///
/// Unfortunately, in Fortran there is no way to disable this behaviour, and so if it
/// is not intended, it's necessary to have a separate assignment statement:
///
/// ```f90
/// subroutine example()
///   integer :: var
///   var = 1
///   print*, var
///   var = var + 1
/// end subroutine example
/// ```
///
/// If the variable's value is intended to be constant, then use the `parameter`
/// attribute instead:
///
/// ```f90
/// subroutine example()
///   integer, parameter :: var = 1
///   print*, var
/// end subroutine example
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct InitialisationInDeclaration {
    name: String,
}

impl Violation for InitialisationInDeclaration {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' is initialised in its declaration and has no explicit `save` or `parameter` attribute")
    }
}

impl AstRule for InitialisationInDeclaration {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Only check in procedures
        node.ancestors().find(|parent| {
            ["function", "subroutine", "module_procedure"].contains(&parent.kind())
        })?;

        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Init in declaration ok for save and parameter
        if declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .any(|attr_name| ["save", "parameter"].contains(&attr_name.to_lowercase().as_str()))
        {
            return None;
        }

        let name = node.child_by_field_name("left")?.to_text(src)?.to_string();
        some_vec![Diagnostic::from_node(Self { name }, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["init_declarator"]
    }
}

/// ## What it does
/// Checks for local pointer variables with implicit `save`
///
/// ## Why is this bad?
/// Initialising procedure local pointer variables in their declaration gives them
/// an implicit `save` attribute: the initialisation is only done on the first call
/// to the procedure, and the pointer retains its associated status on exit.
///
/// However, this associated status makes no guarantee that the target of the pointer
/// is valid or stays allocated between procedure calls - potentially leading to cases
/// where future calls into a procedure will reference an unallocated variable.
///
/// ## Examples
/// For example, this subroutine:
///
/// ```f90
/// subroutine example()
///   integer, target  :: bad
///   integer, pointer :: var => null()
///
///   if (.not.associated(var)) then
///     bad = 1
///     var => bad
///   end if
///
///   print *, var
/// end subroutine example
/// ```
///
/// when called twice
///
/// ```f90
/// call example()
/// call doAnotherBigThing()
/// call example()
/// ```
///
/// will implicitly save the association of `var` with the variable `bad` in the first
/// procedure call, but that variable most likely won't be at the same location in memory in
/// the next call, so the value printed in the second procedure call might not be `1`.
///
/// Adding the `save` attribute makes it clear that this is the intention, however you should
/// also ensure that the target variable is also saved.
///
/// ```f90
/// subroutine example()
///   integer, target,  save :: bad
///   integer, pointer, save :: var => null()
///
///   if (.not.associated(var)) then
///     bad = 1
///     var => bad
///   end if
///
///   print *, var
/// end subroutine example
/// ```
///
/// Unfortunately, in Fortran there is no way to disable this behaviour, and so if it
/// is not intended, it's necessary to have a separate nullification statement before use:
///
/// ```f90
/// subroutine example()
///   integer, target  :: bad
///   integer, pointer :: var
///
///   var => null()  ! or use nullify(var)
///
///   if (.not.associated(var)) then
///     bad = 1
///     var => bad
///   end if
///
///   print *, var
/// end subroutine example
/// ```
///
/// ## References
/// - Metcalf, M., Reid, J. and Cohen, M., 2018, _Modern Fortran Explained:
///   Incorporating Fortran 2018_, Oxford University Press, Section 8.5.3/8.5.4.
/// - Clerman, N. Spector, W., 2012, _Modern Fortran: Style and Usage_, Cambridge
///   University Press, Rule 74, p. 99.
#[derive(ViolationMetadata)]
pub(crate) struct PointerInitialisationInDeclaration {
    name: String,
}

impl Violation for PointerInitialisationInDeclaration {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("Pointer '{name}' is initialized in its declaration and has no explicit `save` attribute")
    }
}

impl AstRule for PointerInitialisationInDeclaration {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Only check in procedures
        node.ancestors().find(|parent| {
            ["function", "subroutine", "module_procedure"].contains(&parent.kind())
        })?;

        let declaration = node
            .ancestors()
            .find(|parent| parent.kind() == "variable_declaration")?;

        // Init in declaration ok for save (in this rule)
        if declaration
            .children_by_field_name("attribute", &mut declaration.walk())
            .filter_map(|attr| attr.to_text(src))
            .any(|attr_name| ["save"].contains(&attr_name.to_lowercase().as_str()))
        {
            return None;
        }

        let var = node.child_by_field_name("left")?;

        // Array syntax on the variable name
        if let Some(arr) = var.child_with_name("identifier") {
            let name = arr.to_text(src)?.to_string();
            return some_vec![Diagnostic::from_node(Self { name }, node)];
        }

        let name = var.to_text(src)?.to_string();
        some_vec![Diagnostic::from_node(Self { name }, node)]
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["pointer_init_declarator"]
    }
}
