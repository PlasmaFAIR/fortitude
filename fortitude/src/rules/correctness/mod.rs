pub mod accessibility_statements;
pub mod assumed_size;
pub mod derived_default_init;
pub mod exit_labels;
pub mod external;
pub mod implicit_kinds;
pub mod implicit_typing;
pub mod init_decls;
pub mod intent;
pub mod kind_suffixes;
pub mod magic_numbers;
pub mod missing_io_specifier;
pub mod select_default;
pub mod trailing_backslash;
pub mod use_statements;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::correctness::exit_labels;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::ImplicitTyping, Path::new("C001.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002.f90"))]
    #[test_case(Rule::ImplicitExternalProcedures, Path::new("C003.f90"))]
    #[test_case(Rule::MissingDefaultCase, Path::new("C011.f90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    #[test_case(Rule::MagicNumberInArraySize, Path::new("C031.f90"))]
    #[test_case(Rule::MagicIoUnit, Path::new("C032.f90"))]
    #[test_case(Rule::MissingActionSpecifier, Path::new("C041.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C051.F90"))]
    #[test_case(Rule::MissingIntent, Path::new("C061.f90"))]
    #[test_case(Rule::AssumedSize, Path::new("C071.f90"))]
    #[test_case(Rule::AssumedSizeCharacterIntent, Path::new("C072.f90"))]
    #[test_case(Rule::InitialisationInDeclaration, Path::new("C081.f90"))]
    #[test_case(Rule::PointerInitialisationInDeclaration, Path::new("C082.f90"))]
    #[test_case(Rule::ExternalProcedure, Path::new("C091.f90"))]
    #[test_case(Rule::ProcedureNotInModule, Path::new("C092.f90"))]
    #[test_case(Rule::MissingDefaultPointerInitalisation, Path::new("C101.f90"))]
    #[test_case(Rule::UseAll, Path::new("C121.f90"))]
    #[test_case(Rule::MissingIntrinsic, Path::new("C122.f90"))]
    #[test_case(Rule::MissingAccessibilityStatement, Path::new("C131.f90"))]
    #[test_case(Rule::DefaultPublicAccessibility, Path::new("C132.f90"))]
    #[test_case(Rule::MissingExitOrCycleLabel, Path::new("C141.f90"))]
    #[test_case(Rule::ExitOrCycleInUnlabelledLoop, Path::new("C142.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::ImplicitTyping, Path::new("C001_ok.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test]
    fn warn_c142_on_nested_loops_only() -> Result<()> {
        let rule_code = Rule::ExitOrCycleInUnlabelledLoop;
        let path = Path::new("C142.f90");
        let snapshot = format!(
            "{}_{}_nested_loops_only",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                exit_unlabelled_loops: exit_labels::settings::Settings {
                    allow_unnested_loops: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
