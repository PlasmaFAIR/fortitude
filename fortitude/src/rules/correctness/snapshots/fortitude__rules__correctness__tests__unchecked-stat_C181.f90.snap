---
source: fortitude/src/rules/correctness/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/correctness/C181.f90:14:20: C181 stat argument 'status' is overwritten before being checked in this scope.
   |
12 |   ! Set stat, but don't check it until after another allocate.
13 |   ! Should raise an error.
14 |   allocate (x(10), stat=status)
   |                    ^^^^^^^^^^^ C181
15 |
16 |   ! Set same stat and check it.
   |

./resources/test/fixtures/correctness/C181.f90:29:20: C181 stat argument 'status' is overwritten before being checked in this scope.
   |
28 |   ! Repeat, but the second allocate is not a direct sibling
29 |   allocate (x(10), stat=status)
   |                    ^^^^^^^^^^^ C181
30 |   if (.true.) then
31 |     allocate (y(10), stat=status)
   |

./resources/test/fixtures/correctness/C181.f90:59:20: C181 stat argument 'status_x' is overwritten before being checked in this scope.
   |
58 |   ! Repeat, but using a custom error handler
59 |   allocate (x(10), stat=status_x)
   |                    ^^^^^^^^^^^^^ C181
60 |   if (.true.) then
61 |     allocate (y(10), stat=status_y)
   |

./resources/test/fixtures/correctness/C181.f90:103:22: C181 stat argument 'status_x' is not checked in this scope.
    |
101 |   ! Stat is not checked in any cases. Should raise for all four.
102 |   if(if_option == 1) then
103 |     allocate (x(10), stat=status_x)
    |                      ^^^^^^^^^^^^^ C181
104 |   else
105 |     allocate (x(30), stat=status_x)
    |

./resources/test/fixtures/correctness/C181.f90:105:22: C181 stat argument 'status_x' is not checked in this scope.
    |
103 |     allocate (x(10), stat=status_x)
104 |   else
105 |     allocate (x(30), stat=status_x)
    |                      ^^^^^^^^^^^^^ C181
106 |   end if
107 |   select case (select_option)
    |

./resources/test/fixtures/correctness/C181.f90:109:22: C181 stat argument 'status_y' is not checked in this scope.
    |
107 |   select case (select_option)
108 |   case(1)
109 |     allocate (y(10), stat=status_y)
    |                      ^^^^^^^^^^^^^ C181
110 |   case default
111 |     allocate (y(20), stat=status_y)
    |

./resources/test/fixtures/correctness/C181.f90:111:22: C181 stat argument 'status_y' is not checked in this scope.
    |
109 |     allocate (y(10), stat=status_y)
110 |   case default
111 |     allocate (y(20), stat=status_y)
    |                      ^^^^^^^^^^^^^ C181
112 |   end select
    |

./resources/test/fixtures/correctness/C181.f90:116:20: C181 stat argument 'status_w' is overwritten before being checked in this scope.
    |
114 |   ! Set stat, but overwrite it before checking it.
115 |   ! Should raise an error.
116 |   allocate (w(10), stat=status_w)
    |                    ^^^^^^^^^^^^^ C181
117 |   status_w = 0
118 |   if (status_w /= 0) then
    |

./resources/test/fixtures/correctness/C181.f90:126:20: C181 stat argument 'status_z' is not checked in this scope.
    |
124 |   ! Set stat, but don't check it until the end of the program.
125 |   ! Should raise an error.
126 |   allocate (z(10), stat=status_z)
    |                    ^^^^^^^^^^^^^ C181
127 |
128 |   print *, "Something between allocate and the end of the program"
    |

./resources/test/fixtures/correctness/C181.f90:146:22: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
144 |     real(8), allocatable :: x(:)
145 |     logical :: file_exist
146 |     inquire(unit=10, iostat=status, exist=file_exist)
    |                      ^^^^^^^^^^^^^ C181
147 |     open (file="file.txt", unit=10, iostat=status, asynchronous="YES")
148 |     write(10, *, iostat=status) line
    |

./resources/test/fixtures/correctness/C181.f90:147:37: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
145 |     logical :: file_exist
146 |     inquire(unit=10, iostat=status, exist=file_exist)
147 |     open (file="file.txt", unit=10, iostat=status, asynchronous="YES")
    |                                     ^^^^^^^^^^^^^ C181
148 |     write(10, *, iostat=status) line
149 |     flush(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:148:18: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
146 |     inquire(unit=10, iostat=status, exist=file_exist)
147 |     open (file="file.txt", unit=10, iostat=status, asynchronous="YES")
148 |     write(10, *, iostat=status) line
    |                  ^^^^^^^^^^^^^ C181
149 |     flush(10, iostat=status)
150 |     backspace(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:149:15: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
147 |     open (file="file.txt", unit=10, iostat=status, asynchronous="YES")
148 |     write(10, *, iostat=status) line
149 |     flush(10, iostat=status)
    |               ^^^^^^^^^^^^^ C181
150 |     backspace(10, iostat=status)
151 |     rewind(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:150:19: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
148 |     write(10, *, iostat=status) line
149 |     flush(10, iostat=status)
150 |     backspace(10, iostat=status)
    |                   ^^^^^^^^^^^^^ C181
151 |     rewind(10, iostat=status)
152 |     read(10, *, iostat=status, asynchronous="YES") line
    |

./resources/test/fixtures/correctness/C181.f90:151:16: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
149 |     flush(10, iostat=status)
150 |     backspace(10, iostat=status)
151 |     rewind(10, iostat=status)
    |                ^^^^^^^^^^^^^ C181
152 |     read(10, *, iostat=status, asynchronous="YES") line
153 |     wait(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:152:17: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
150 |     backspace(10, iostat=status)
151 |     rewind(10, iostat=status)
152 |     read(10, *, iostat=status, asynchronous="YES") line
    |                 ^^^^^^^^^^^^^ C181
153 |     wait(10, iostat=status)
154 |     endfile(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:153:14: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
151 |     rewind(10, iostat=status)
152 |     read(10, *, iostat=status, asynchronous="YES") line
153 |     wait(10, iostat=status)
    |              ^^^^^^^^^^^^^ C181
154 |     endfile(10, iostat=status)
155 |     close(10, iostat=status)
    |

./resources/test/fixtures/correctness/C181.f90:154:17: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
152 |     read(10, *, iostat=status, asynchronous="YES") line
153 |     wait(10, iostat=status)
154 |     endfile(10, iostat=status)
    |                 ^^^^^^^^^^^^^ C181
155 |     close(10, iostat=status)
156 |     ! Something other than IO statement, reuse status
    |

./resources/test/fixtures/correctness/C181.f90:155:15: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
153 |     wait(10, iostat=status)
154 |     endfile(10, iostat=status)
155 |     close(10, iostat=status)
    |               ^^^^^^^^^^^^^ C181
156 |     ! Something other than IO statement, reuse status
157 |     allocate(x(10), stat=status)
    |

./resources/test/fixtures/correctness/C181.f90:157:21: C181 stat argument 'status' is overwritten before being checked in this scope.
    |
155 |     close(10, iostat=status)
156 |     ! Something other than IO statement, reuse status
157 |     allocate(x(10), stat=status)
    |                     ^^^^^^^^^^^ C181
158 |     deallocate(x, stat=status)
159 |   end subroutine io
    |

./resources/test/fixtures/correctness/C181.f90:163:37: C181 stat argument 'status' is not checked in this scope.
    |
161 |   subroutine cmd_line()
162 |     integer :: status
163 |     call execute_command_line("ls", cmdstat=status)
    |                                     ^^^^^^^^^^^^^^ C181
164 |   end subroutine cmd_line
    |
